### PR TITLE
Starwood Scepter

### PR DESCRIPTION
**Pull request by SkippZzMods**
Implements the starwood scepter summon weapon and also the obtainment item for it

https://trello.com/c/5CtcZLD3/20-starwood-scepter-45

Should be finished, its pretty balanced, and its got obtainment. Multiplayer sync has not been tested yet, and will probably need some stuff there considering all of the projectile instances that are being passed.

### CODE REVIEW

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+namespace StarlightRiver.Content.Items.Starwood
+{
+	public class StarwoodScepterPool : LootPool
+	{
+		public override void AddLoot()
+		{
+			AddItem(ModContent.ItemType<StarwoodScepter>(), ChestRegionFlags.Surface, 0.5f, 1, true, 2);
+		}
+	}
+	public class DormantScepter : QuickMaterial
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:19]: While this is pretty small, its porbably best this item have it's own file

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+					return Owner.Center + new Vector2(-25 * Owner.direction, -50) + new Vector2(-35 * Owner.direction * otherProj.minionPos, 0f) + new Vector2(-MathHelper.Lerp(5f, 15f, Utils.Clamp((float)Math.Sin(lifetime * 0.015f), 0, 1))).RotatedBy(MathHelper.ToRadians(lifetime));
+			}
+		}
+
+		public StarwoodScepterSummonSplit OtherStar => otherProj.ModProjectile as StarwoodScepterSummonSplit;
+		public Projectile EmpoweredStar => Main.projectile.Where(p => p.active && p.type == ModContent.ProjectileType<StarwoodScepterSummonEmpowered>() && (p.ModProjectile as StarwoodScepterSummonEmpowered).Children.Contains(Projectile.whoAmI)).FirstOrDefault();
+		public bool HasEmpoweredStar => EmpoweredStar != null;
+		public bool IsEmpowered => Owner.GetModPlayer<StarlightPlayer>().empowered;
+		public bool FoundTarget => Target != null;
+		public ref float TargetWhoAmI => ref Projectile.ai[0];
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:112]: This is likely a poor candidate for an AI field, as whoAmI's change across clients

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+	}
+
+	public class StarwoodScepterSummonSplit : ModProjectile
+	{
+		public int lifetime;
+		public int empowermentTimer;
+		public int downtimeTimer;
+		public Projectile otherProj;
+		public Vector2 rotationalVelocity = Vector2.Zero;
+		public NPC MinionTarget
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:85]: The property with getters/setters should appear after autoproperties.

Additionally, linebreaks seperating fields from properties are neccisary, and the properties should likely be grouped and have linebreaks with reasonable relationships to each other

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+
+			if (Main.rand.NextBool(10) && Owner.HasMinionAttackTargetNPC && Owner.MinionAttackTargetNPC == target.whoAmI && hasTag)
+			{
+				if (Projectile.owner == Main.myPlayer)
+				{
+					int newItem = Item.NewItem(target.GetSource_Loot(), Projectile.Hitbox, ItemID.Star); // Create a new item in the world.
+					Main.item[newItem].noGrabDelay = 0; // Set the new item to be able to be picked up instantly
+
+					// Here we need to make sure the item is synced in multiplayer games.
+					if (Main.netMode == NetmodeID.MultiplayerClient && newItem >= 0)
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:206]: single line conditional should be inlined

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+
+			return false;
+		}
+
+		internal void DefaultBehavior()
+		{
+			if (empowermentTimer > 0)
+				empowermentTimer = 0;
+
+			if (!IsParent)
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:247]: single line conditional should be inlined

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+			{
+				DoIdleMovement();
+
+				if (Main.rand.NextBool(90))
+					Dust.NewDustPerfect(Projectile.Center, ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2Circular(2f, 2f), 0, Main.rand.NextBool() ? new Color(255, 200, 0) : new Color(255, 150, 0), 0.35f);
+
+				if (IsParent)
+				{
+					NPC target = FindTarget();
+					if (target != default)
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:274]: single line condtional should be inlined

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+
+				if (empowermentTimer < 25)
+				{
+					Projectile.velocity *= 0.94f;
+					Projectile.rotation += Projectile.velocity.Length() * 0.02f;
+					rotationalVelocity = Projectile.rotation.ToRotationVector2();
+				}
+				else
+				{
+					if (empowermentTimer == 25)
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:347]: single line conditional should be inlined

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 8]
```diff
@@ -0,0 +1,741 @@
...
+		{
+			return Main.npc.Where(n => n.CanBeChasedBy() && n.Distance(Projectile.Center) < 1000f).OrderBy(n => n.Distance(Projectile.Center)).FirstOrDefault();
+		}
+	}
+
+	public class StarwoodScepterSummonEmpowered : ModProjectile
+	{
+		public int downtimeTimer;
+		public int[] Children = new int[2];
+		public Projectile[] ProjectileChildren => new Projectile[2] { Main.projectile[Children[0]], Main.projectile[Children[1]] };
```
Comment (ScalarVector1) [inline Content/Items/Starwood/Weapons.StarwoodScepter.cs:448]: same as above, please seperate and organize properties appropriately

**Review by ScalarVector1 — CHANGES_REQUESTED** [review]
Looks fine functionally but needs code quality/styling fixes.

Additionally, XML comments on the AI sub-methods would be very nice to have.

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs (hunk 2 above)
Comment (SkippZzMods) [reply Content/Items/Starwood/Weapons.StarwoodScepter.cs:112]: wait even npc ones

Comment (SkippZzMods) [reply Content/Items/Starwood/Weapons.StarwoodScepter.cs:112]: god terrarias netcode is so fucking atrocious

File: Content/Items/Starwood/Weapons.StarwoodScepter.cs [hunk 9]
```diff
@@ -0,0 +1,773 @@
...
+using StarlightRiver.Core.Systems.ChestLootSystem;
+using System;
+using System.Linq;
+using Terraria.Audio;
+using Terraria.DataStructures;
+using Terraria.ID;
+
+namespace StarlightRiver.Content.Items.Starwood
+{
+	public class StarwoodScepterPool : LootPool
```
[the remark this answers is not in the document]
Comment (SkippZzMods) [reply Content/Items/Starwood/Weapons.StarwoodScepter.cs:12]: its item specific so i opted to keep it in the file. I could make a StarwoodLootPool in the folder but, i doubt anything will use it. Whichever seems better to you

[the remark this answers is not in the document]
Comment (ScalarVector1) [reply Content/Items/Starwood/Weapons.StarwoodScepter.cs:12]: I mean to have this in the same file as the dormant scepter as the loot is the material scepter, not the crafted one

**Review by ScalarVector1 — APPROVED** [review]

**Pull request merged**